Update cibuildwheel to 2.15.0

richli opened the pull request:
The motivation for this update is to start building wheels for Python 3.12 by default.

richli commented:
[Changelog for 2.15.0](https://github.com/pypa/cibuildwheel/blob/v2.15.0/docs/changelog.md)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Unidata/cftime?pullRequest=309) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Unidata/cftime?pullRequest=309) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Unidata/cftime?pullRequest=309) it.</sub>

richli commented:
Oh, maybe this was jumping the gun a bit, numpy wheels for Python 3.12 aren't out yet, but it looks numpy 1.26 will address that.

richli commented:
Based on [this comment](https://github.com/numpy/numpy/issues/23808#issuecomment-1678677094) I updated the build-time and run-time dependencies for Python 3.12 so that it can use the pre-release numpy version `1.26.0b1`.

The `cftime` wheel now builds for Python 3.12, but during the test phase when installing from `requirements-dev.txt`, it cannot yet install `twine` so I've disabled that. Presumably this is only temporary and that requirement specifier can be removed later.

richli commented:
It turns out that reusing `requirements.txt` and `requirements-dev.txt` to specify a conda environment doesn't work with the [requirement specifiers](https://pip.pypa.io/en/latest/reference/requirement-specifiers/) that were added. So I updated the CI config to not use them.

In the process, I updated the `mamba-org/provision-with-micromamba` action, which was deprecated, to its replacement, `mamba-org/setup-micromamba`, and updated a couple of things to bring the CI up-to-date. So hopefully everything works again.
